Section one

dsamojlenko opened the pull request:
- Add introductory text to the Start page
- Add Section 1 (Personal Information) of the form, with validation schema
- Create custom validation helpers (in utils/custom):
  - **isValidDate** to validate any date format (just check if it's a valid date)
  - **validateStringFormat** to validate any string against a regex for formatting (ie, YYYY-MM-DD for dates)
  - **validateTelephoneFormat** uses validateStringFormat to validate phone number format ###-###-####
  - **validateDateFormat** uses validateStringFormat to validate date format YYYY-MM-DD
  - **validateSINFormat** uses validateStringFormat to validate format ### ### ###
  - **validateSIN** validates the SIN using Luhn Algorithm
- Fix a minor bug in radios template

**For testing**: [Generate a fake SIN](https://www.fakenamegenerator.com/social-insurance-number.php) that passes the Luhn test

dsamojlenko committed:
add intro txt from pdf

dsamojlenko committed:
add section one content

dsamojlenko committed:
simplify date validation

dsamojlenko committed:
use index as value for radios

dsamojlenko committed:
add validation

dsamojlenko committed:
revert to original - moved custom into separate file

dsamojlenko committed:
add custom validation helpers

dsamojlenko committed:
run multiple custom validation rules

dsamojlenko committed:
really revert

dsamojlenko committed:
validate phone number format

dsamojlenko committed:
create custom validators to cleanup schema

dsamojlenko committed:
rename method for consistency

dsamojlenko committed:
fix linting errors

CalvinRodo commented on routes/personal/schema.js:
```diff
@@ -1,56 +1,74 @@
 
 const Schema = {
-  fullname: {
+  social: {
```
SIN uses the Luhn Algorithm for validation https://www.npmjs.com/search?q=luhn if you want to implement a custom validation for it

dsamojlenko replied:
added

CalvinRodo commented on routes/personal/schema.js:
```diff
@@ -1,56 +1,74 @@
+  last_name: {
+    isLength: {
+      errorMessage: 'Last name is required',
+      options: { min: 3, max: 40 },
```
Not we should put in minimums for last names since a fairly common last name is ng in some parts of the world, and so they would not be able to enter the data.

CalvinRodo replied:
Also validating names is hard

dsamojlenko replied:
agreed. I'm going to set length to min:1 just to make the fields required.

dsamojlenko committed:
add alternative phone validation

dsamojlenko committed:
validate SIN

dsamojlenko committed:
make first_name and last_name required (don't validate length)

dsamojlenko committed:
Validate SIN using Luhn Algorithm

dsamojlenko committed:
fix linting errors

CalvinRodo commented on utils/custom/validate.helpers.js:
```diff
@@ -0,0 +1,58 @@
+
+const validateSINFormat = string => {
+    // validates format: ### ### ###
+    const format = /^(\d{3} \d{3} \d{3})$/
```
should we care if they put spaces in the sin the following regex should take care of this: `/^(\d{3} ?\d{3} ?\d{3})$/` although if we are going to replace this with a masked input this becomes moot.

CalvinRodo replied:
alternatively we can just strip out white space

dsamojlenko replied:
I'll modify it to pass with or without spaces ... maybe we can add field masking later for this and telephone fields.

CalvinRodo commented on utils/custom/validate.helpers.js:
```diff
@@ -0,0 +1,58 @@
+const isValidDate = dateString => {
+  return !isNaN(Date.parse(dateString));
```
According to MDN's article on [Date.parse](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/parse)

>It is not recommended to use Date.parse as until ES5, parsing of strings was entirely implementation dependent. There are still many differences in how different hosts parse date strings, therefore date strings should be manually parsed (a library can help if many different formats are to be accommodated).

So we'd probably want to use a library here maybe something like luxon or date-fns that has immutable dates (so not moment.js) 

https://blog.bitsrc.io/9-javascript-date-time-libraries-for-2018-12d82f37872d

CalvinRodo replied:
Also some weird functionality for this it'll flag the following as a valid date `0001-01-101` and last I checked it wasn't although I'm open to the fact that when it comes to dates in the year 0001 that I'm not exactly an expert.

dsamojlenko replied:
interesting... ok I'll swap it out for some other code I had been testing before I found this.

dsamojlenko committed:
add some tests

CalvinRodo commented on utils/custom/validate.helpers.js:
```diff
@@ -0,0 +1,58 @@
+
+const validateDateFormat = string => {
+    // validates format: YYYY-MM-DD
+    const format = /(\d{4})-(\d{2})-(\d{2})/
```
should this regex not capture the beginning and end of the string like the others so: 
`/^(\d{4})-(\d{2})-(\d{2})$/`

CalvinRodo replied:
If fixed this in my push

CalvinRodo committed:
Adding unit tests for validation code

Fixed the Date Regex so that it now captures the beginning and the end
of the string.

CalvinRodo committed:
Merging changes and adding some tests

CalvinRodo commented:
ack I guess I don't have code signing setup, @dsamojlenko can you resign all commits here 😦 

https://superuser.com/questions/397149/can-you-gpg-sign-old-commits sorry about that.

dsamojlenko reviewed (commented):
@CalvinRodo 

I think yours are signed - you're just showing as Unverified because you have to copy your gpg public key to your github profile here: https://github.com/settings/keys

dsamojlenko committed:
optionally allow spaces in SIN

dsamojlenko committed:
display valid example SIN for easier testing

dsamojlenko committed:
better date parsing and some refactoring

CalvinRodo reviewed (approved):
:+1:
